core/client/fs/src/main/java/alluxio/client/block/stream/NaiveSharedGrpcDataReader.java

```diff
@@ -0,0 +1,175 @@
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.client.block.stream;
+
+import alluxio.client.file.FileSystemContext;
+import alluxio.conf.AlluxioConfiguration;
+import alluxio.conf.PropertyKey;
+import alluxio.grpc.DataMessage;
+import alluxio.grpc.ReadRequest;
+import alluxio.grpc.ReadResponse;
+import alluxio.grpc.ReadResponseMarshaller;
+import alluxio.network.protocol.databuffer.DataBuffer;
+import alluxio.network.protocol.databuffer.NioDataBuffer;
+import alluxio.resource.CloseableResource;
+import alluxio.resource.LockResource;
+import alluxio.wire.WorkerNetAddress;
+
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import com.google.common.base.MoreObjects;
+import com.google.common.base.Preconditions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * A gRPC data reader that streams a region from gRPC data server.
+ *
+ * NaiveSharedGrpcDataReader is a POC to solve multi-process-read perf issue.
+ * It follows GrpcDataReader protocol.
+ * It takes strong assumption:
+ *    Parallel read to the same file happens on the same time, so that read request is 
+ *    serialized by kernel
+ */
+@NotThreadSafe
+public final class NaiveSharedGrpcDataReader implements DataReader {
+  private static final Logger LOG = LoggerFactory.getLogger(NaiveSharedGrpcDataReader.class);
+
+  private final NaiveCachedGrpcDataReader mCachedDataReader;
+  private final long mBlockId;
+
+  private long mChunkSize;
+  private long mBlockSize;
+
+  /** The next pos to read. */
+  private long mPosToRead;
+
+  private final static ReentrantReadWriteLock mBlockLocks = new ReentrantReadWriteLock();
+  private final static Map<Long, NaiveCachedGrpcDataReader> mBlockReaders = new HashMap<>();
+
+  /**
+   * Creates an instance of {@link NaiveSharedGrpcDataReader}.
+   *
+   * @param context the file system context
+   * @param address the data server address
+   * @param readRequest the read request
+   */
+  private NaiveSharedGrpcDataReader(FileSystemContext context, WorkerNetAddress address,
+      ReadRequest readRequest) throws IOException {
+    mChunkSize = readRequest.getChunkSize();
+    mPosToRead = readRequest.getOffset();
+    mBlockSize = readRequest.getLength() + readRequest.getOffset();
+    try (LockResource r1 = new LockResource(mBlockLocks.writeLock())) {
```

Comment:
I actually think this part of looking up and returning cached reader should go to factory rather than constructor here

Reply:
Moved